File: 2016/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Room = tuple[str, int, str]


def solution(content: list[str]) -> Any:
    return sum(
        num
        for name, num, check in (parse(line) for line in content)
        if check == checksum(name)
    )


def checksum(name: str):
    def helper(count: dict[str, int], line: list[str]):
        if not line:
            return "".join(
                key
                for key, _ in sorted(
                    sorted(count.items(), key=lambda item: item[0]),
                    key=lambda item: item[1],
                    reverse=True,
                )
            )[:5]
        char, *rest = line
        count[char] = 1 + count.get(char, 0)
        return helper(count, rest)

    return helper({}, list(name))


def parse(line: str) -> Room:
    match = re.match(r"([a-z\-]+)(\d+)\[([a-z]+)\]", line)
    if match is None:
        raise ValueError(f"Invalid line: {line}")
    return (match.group(1).replace("-", ""), int(match.group(2)), match.group(3))


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2016/day04-1.hs
import Data.Function (on)
import Data.List (sortBy)
import Distribution.Utils.Generic (sndOf3)

type Room = (String, Int, String)

solution :: String -> Int
solution = sum . map sndOf3 . filter (\(name, _, chs) -> checksum name == chs) . map parse . lines

third :: (a, b, c) -> c
third (_, _, a) = a

checksum :: String -> String
checksum = map snd . take 5 . sortBy (flip compare `on` fst) . sortBy (compare `on` snd) . helper []
  where
    helper :: [(Int, Char)] -> String -> [(Int, Char)]
    helper count [] = count
    helper count ('-' : ys) = helper count ys
    helper count (' ' : ys) = helper count ys
    helper count (y : ys)
      | y `elem` map snd count = helper ([(c + 1, z) | (c, z) <- count, z == y] ++ filter ((/= y) . snd) count) ys
      | otherwise = helper ((1, y) : count) ys

parse :: String -> Room
parse xs = (unwords . init . init $ parts, read . head . tail . reverse $ parts, last parts)
  where
    parts = splitOn "-[]" xs

splitOn :: String -> String -> [String]
splitOn cs xs = helper [""] cs xs
  where
    helper o c "" = o
    helper o c (x : "")
      | x `elem` cs = o
      | otherwise = init o ++ [last o ++ [x]]
    helper o c (x : xs)
      | x `elem` cs = helper (o ++ [""]) cs xs
      | otherwise = helper (init o ++ [last o ++ [x]]) cs xs

main :: IO ()
main = do readFile "day04.dat" >>= print . solution
